### DISCUSSION TITLE
why assertion will fail in BufferStream if const string size shorter than buffer size

### DISCUSSION THREAD

**Discussion by LukyGuyLucky** [c3lang/c3c#1274] [Q&A]
[https://github.com/c3lang/c3c/blob/master/test/unit/stdlib/io/bufferstream.c3](url)

```
const DATA="a short one";
fn void! writebuffer_large()
{
	ByteWriter out;
	out.temp_init();
	char[9] buf;
	WriteBuffer write_buf;
	write_buf.init(&out,buf[..]);
	
	usz n=write_buf.write(DATA)!;
	
	if(n!=DATA.len)
	{
		libc::printf("large write failed #1: got %d;want %d.",n,DATA.len);
		
	}
	String got=out.str_view();
	if(got!=DATA)
	{
		io::printfn("large write failed #2:got %d,want %d",got,DATA);//---->will trigger this line if DATA.len is shorter than char[16] buf size?
	}
}
```

It would be grateful if a FileStream library or example is provided.Manipulating file in a stream way is desired.
Thanks in advance.

It is great to know io::print family and c printf family has been integrated and simplified. I was wondering is there any difference between io::printf and lic::printf?And also it is great more STL functions and methods have been integrated into the handbook.But I failed to found int system(char* command) has been listed in(or put it in another way,how to keep the terminal window open to wait user input before the program exits?).

But unlucky since 0.60 I failed to compile cforms  [https://github.com/kcvinker/cforms](url)  /sqlite3/raylib examples discussed before.Would you please take a look at them and provide a workable example(s) when you are available?

Appreciated.

**Comment by lerno** [comment]
In this case, look at what you're doing:

```c
// For C:
libc::printf("large write failed #1: got %d;want %d.",n,DATA.len);
```
This is *almost* fine, you're sending n (which is a number) and DATA.len (also a number) to printf. The problem is that if you use libc printf you need to use libc printf rules, so for an `usz` you need to use `%llu` rather than `%d`.

The second you try to use C3:
```c
io::printfn("large write failed #2:got %d,want %d",got,DATA)
```
Now you are passing `got` – which is a `String` and `DATA`, which is also a `String` to a %d parameter. %d parameters are for integers, so this is why you get this:

```
ERROR: 'Unexpected fault 'INVALID_ARGUMENT_TYPE' was unwrapped!'
  in std.core.builtin.panicf 
```

As the error message says, there's an invalid argument type which has been passed to printf.

If you change it to %s, or if you instead get the length (like in the libc::printf) it will just work.

**Reply by lerno** [reply]
In addition, the reason why this fails is that you have not invoked write_buf.flush() before you're checking the output.
